### ISSUE TITLE
S4018 FP: when type parameter is used in return value.

### ISSUE THREAD

**Issue by dmitry-golovinov-sonarsource** [SonarSource/sonar-dotnet#4548]
### Description

It makes sense to use type parameter only in return value, when dealing with generics.

### Repro steps

```
class Foo
{
        public TKey MyMethod_08<TKey>() { return default(TKey); } // Noncompliant FP

        public static T MyMethod_23<T>(string value) // Noncompliant FP
        {
            return (T) Convert.ChangeType(value, typeof(T));
        }
}
```

### Expected behavior

Issue should not be raised.

### Actual behavior

Issue is being raised.

### Related information

* C#/VB.NET Plugins version 8.22

**Comment by andrei-epure-sonarsource** [comment]
related : #3145

**Comment by martin-strecker-sonarsource** [comment]
Related https://github.com/SonarSource/sonar-dotnet/issues/6529

**Comment by zsolt-kolbay-sonarsource** [comment]
The rule is about type inference, and the type parameter being the return type or being part of it doesn't improve that.
This fix would therefore contradict the rule itself, so after some discussion, we decided to close it as Won't Fix.
Instead, we're going to make the rule's message clear to reflect that the aim is to be able to use type inference when invoking the method (see [this issue](https://github.com/SonarSource/sonar-dotnet/issues/6636)).
